File: python/tests/test_stream_compaction.py
# Copyright (c) 2025, NVIDIA CORPORATION
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
import pyarrow as pa
import pytest

from legate_dataframe import LogicalColumn, LogicalTable
from legate_dataframe.lib.stream_compaction import apply_boolean_mask, distinct
from legate_dataframe.testing import (
    assert_arrow_table_equal,
    assert_matches_polars,
    get_pyarrow_column_set,
    std_dataframe_set,
)


@pytest.mark.parametrize("df", std_dataframe_set())
def test_apply_boolean_mask_basic(df: pa.Table):
    lg_df = LogicalTable.from_arrow(df)
    np.random.seed(0)
    mask = pa.array(np.random.randint(0, 2, size=len(df)).astype(bool))
    lg_mask = LogicalColumn.from_arrow(mask)

    res = apply_boolean_mask(lg_df, lg_mask)
    expect = df.filter(mask)

    assert_arrow_table_equal(res.to_arrow(), expect)


@pytest.mark.parametrize("df", std_dataframe_set())
def test_apply_boolean_mask_nulls(df: pa.Table):
    # Similar to `test_apply_boolean_mask`, but cover a nullable column
    lg_df = LogicalTable.from_arrow(df)

    np.random.seed(0)
    mask_values = np.random.randint(0, 2, size=len(df)).astype(bool)
    mask_mask = np.random.randint(0, 2, size=len(df)).astype(bool)
    mask = pa.array(mask_values, mask=mask_mask)
    lg_mask = LogicalColumn.from_arrow(mask)

    res = apply_boolean_mask(lg_df, lg_mask)
    expect = df.filter(mask)

    assert_arrow_table_equal(res.to_arrow(), expect)


@pytest.mark.parametrize(
    "bad_mask",
    [
        pa.array([1, 0, 1, 0]),  # not boolean
        # wrong length, but as of writing not caught before at/task launch:
        pytest.param(
            pa.array([True, False, False, True, False]), marks=pytest.mark.skip
        ),
    ],
)
def test_apply_boolean_mask_errors(bad_mask):
    df = pa.table({"a": [1, 2, 3, 4]})

    lg_df = LogicalTable.from_arrow(df)
    bad_mask = LogicalColumn.from_arrow(bad_mask)

    with pytest.raises(ValueError):
        apply_boolean_mask(lg_df, bad_mask)


@pytest.mark.parametrize(
    "keys, high_cardinality",
    [
        (["b", "a"], True),
        (["a", "b"], False),
        (["b", "c"], False),
        (None, False),
    ],
)
def test_distinct_basic(keys, high_cardinality):
    df = pa.table(
        {"a": [1, 1, 2, 3, 4], "b": [2, 2, 4, 4, None], "c": [10, 10, 20, 30, 40]}
    )
    lg_df = LogicalTable.from_arrow(df)
    res = distinct(lg_df, keys, high_cardinality=high_cardinality)
    expect = pa.table({"a": [1, 2, 3, 4], "b": [2, 4, 4, None], "c": [10, 20, 30, 40]})

    # Result has no defined order, so need to sort before comparing
    sort_by = [("a", "ascending"), ("b", "ascending"), ("c", "ascending")]
    assert_arrow_table_equal(res.to_arrow().sort_by(sort_by), expect.sort_by(sort_by))


@pytest.mark.parametrize(
    "arrow_column", get_pyarrow_column_set(["int32", "float32", "int64"])
)
def test_column_filter_polars(arrow_column):
    pl = pytest.importorskip("polars")

    q = pl.DataFrame({"a": arrow_column}).lazy()
    q = q.filter(pl.col("a") > 0.5)

    assert_matches_polars(q)


@pytest.mark.parametrize("dtype", ["float64", "int32", "int64"])
@pytest.mark.parametrize("keys", [["a", "b"], ["a"], ["b"], None])
def test_polars_distinct(dtype, keys):
    pl = pytest.importorskip("polars")

    mask = np.random.randint(2, size=10_000, dtype=bool)
    a = np.random.randint(100, size=10_000).astype(dtype)
    if dtype == "float64":
        a[[200, 500, 600]] = np.nan  # see how NaNs are handled.

    b = -a  # Don't deal with duplicates, since we can get any value back
    b[mask] = 1

    q = pl.DataFrame(
        {
            "a": pa.array(a, mask=mask),
            "b": b,
        }
    ).lazy()

    # NOTE(seberg): Descending sort because NaN sorting behaves differently for arrow
    # (That may need working around in sort; as of 2025-10.)
    assert_matches_polars(q.unique(keys).sort(["a", "b"], descending=True))

File: cpp/tests/test_stream_compaction.cpp
/*
 * Copyright (c) 2024-2025, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <arrow/compute/api.h>
#include <gtest/gtest.h>
#include <legate_dataframe/stream_compaction.hpp>

using namespace legate::dataframe;

TEST(StreamCompactionTest, ApplyBooleanMask)
{
  LogicalColumn col_0(std::vector<int32_t>{5, 4, 3, 1, 2, 0});
  LogicalColumn col_1(std::vector<std::string>{"this", "is", "a", "string", "column", "!"});
  LogicalColumn col_2(std::vector<double>{0, 1, 2, 3, 4, 5});

  LogicalColumn boolean_mask(std::vector<bool>{true, false, true, true, false, false});

  LogicalTable tbl{{col_0, col_1, col_2}, {"a", "b", "c"}};

  auto result   = apply_boolean_mask(tbl, boolean_mask);
  auto expected = ARROW_RESULT(arrow::compute::CallFunction(
                                 "filter", {tbl.get_arrow(), boolean_mask.get_arrow()}))
                    .table();
  EXPECT_TRUE(result.get_arrow()->Equals(*expected));

  // Additionally, check that a null mask is honored by legate-dataframe
  LogicalColumn boolean_mask_nulls(std::vector<bool>{true, false, true, true, false, false},
                                   {true, true, true, false, false, true});

  result   = apply_boolean_mask(tbl, boolean_mask_nulls);
  expected = ARROW_RESULT(arrow::compute::CallFunction(
                            "filter", {tbl.get_arrow(), boolean_mask_nulls.get_arrow()}))
               .table();
  EXPECT_TRUE(result.get_arrow()->Equals(*expected));
}

TEST(StreamCompactionTest, Distinct)
{
  LogicalColumn col_0(std::vector<int32_t>{0, 0, 0, 1, 1, 1},
                      {false, false, false, true, true, true});
  LogicalColumn col_1(std::vector<std::string>{"this", "this", "string", "string", "col", "col"});
  LogicalColumn col_2(std::vector<double>{0, 0, 2, 3, 4, 4});

  LogicalTable tbl{{col_0, col_1, col_2}, {"a", "b", "c"}};

  auto result = distinct(tbl, {"a", "b"});

  // Hardcoded expected result (we have no simple arrow unique call).
  LogicalColumn col_0_exp(std::vector<int32_t>{0, 0, 1, 1}, {false, false, true, true});
  LogicalColumn col_1_exp(std::vector<std::string>{"this", "string", "string", "col"});
  LogicalColumn col_2_exp(std::vector<double>{0, 2, 3, 4});
  LogicalTable expected{{col_0_exp, col_1_exp, col_2_exp}, {"a", "b", "c"}};
  auto expected_arrow = expected.get_arrow();

  EXPECT_TRUE(result.get_arrow()->Equals(*expected_arrow));
}
